File: LeetCode/Array/Golang/Leetcode904-fruit-into-baskets.go
package sbgo

func totalFruit(fruits []int) int {
  mp,n,res := map[int]int{}, len(fruits), 0
  max := func(i,j int) int {
    if i < j {
      return j
    }
    return i
  }
  for i, j, s := 0, 0, 0;i < n;i++ {
    mp[fruits[i]]++
    if mp[fruits[i]] == 1 {
      s++
    }
    for s > 2 {
      mp[fruits[j]]--
      if mp[fruits[j]] == 0 {
        s--
      }
      j++
    }
    res = max(res, i - j + 1)
  }
  return res
}

File: LeetCode/Array/Python/Leetcode904-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        mp,res,n = defaultdict(int), 0, len(fruits)
        j,s = 0,0
        for i in range(n):
            mp[fruits[i]] += 1
            if mp[fruits[i]] == 1: s += 1
            while s > 2:
                mp[fruits[j]] -= 1
                if mp[fruits[j]] == 0:
                    s -= 1
                j += 1
            res = max(res, i - j + 1)
        return res
